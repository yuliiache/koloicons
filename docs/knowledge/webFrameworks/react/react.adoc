== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







