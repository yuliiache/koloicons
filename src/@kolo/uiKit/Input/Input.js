import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import styles from './Input.module.scss';

export const Input = ({id, label, placeholder, isDisabled = false, errorMessage, hint, value, onChange}) => {
  const inputClasses = classNames(styles.inputGroupInput, {
    [styles.invalid]: errorMessage,
  });

  const handleChange = (e) => {
    onChange(e.target.value);
  };

  return (
    <div className={styles.inputGroup}>
      {label && (
        <label
          className={styles.inputGroupLabel}
          htmlFor={id}
        >
          {label}
        </label>
      )}
      {hint && <span className={styles.inputGroupHint}>{hint}</span>}
      <input
        className={inputClasses}
        type="text"
        id={id}
        placeholder={placeholder}
        disabled={isDisabled}
        value={value}
        onChange={handleChange}
      />
      {errorMessage && <span className={styles.inputGroupError}>{errorMessage}</span>}
    </div>
  );
};

Input.propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.string,
  placeholder: PropTypes.string,
  isDisabled: PropTypes.bool,
  errorMessage: PropTypes.string,
  hint: PropTypes.string,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
};
