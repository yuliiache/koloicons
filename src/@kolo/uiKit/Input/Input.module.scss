@import "../../design";

.inputGroup {
  position: relative;
  display: flex;
  flex-wrap: wrap;
  padding-bottom: 16px;
  max-width: 50rem;
  margin: auto;

  &Hint {
    margin-left: auto;
    font-size: 0.6rem;
    line-height: 1.5;
    color: $GreyColor;
  }

  &Label {
    display: block;
    margin-bottom: 8px;
    font-weight: 500;
    font-size: 12px;
    line-height: 1.33;
    color: $secFontColor;
    cursor: pointer;
  }

  &Input {
    height: 56px;
    box-sizing: border-box;
    width: 100%;
    padding: 12px 16px;
    font-size: 0.8rem;
    color: $secFontColor;
    font-family: $mainFontFamily;
    line-height: 1.56;
    background: linear-gradient(to bottom, $IconInvert, $IconInvert);
    border: solid 1px $ActiveBorderColor;
    border-radius: 4px;
    outline: none;
    box-shadow: inset 0 2px 0 0 $IconInvert, inset 0 24px 35px -4px $ShadowColor;
    transition: box-shadow 0.3s ease-in-out;

    &::-webkit-input-placeholder {
      color: $lighthGrey;
      font-family: $mainFontFamily;
      font-size: 0.8rem;
      line-height: 1.5;
    }

    &:disabled {
      background: linear-gradient(to bottom, $ShadowColor, $BlackColor 65%),
      linear-gradient(to bottom, $LightBG, $LightBG);
    }

    &:not(:disabled) {
      &:hover, &:active {
        box-shadow: inset 0 2px 0 0 $IconInvert, inset 0 24px 35px -4px $ActiveInputColor;
      }

      &:active {
        border: solid 1px $ActiveBorderColor;
        box-shadow: inset 0 2px 0 0 $IconInvert, inset 0 -2px 0 0 $BrandColor;
      }

      &:focus {
        box-shadow: inset 0 2px 0 0 $IconInvert, inset 0 24px 35px -4px $ActiveInputColor,
        inset 0 -2px 0 0 $BrandColor, 0 0 0 4px $FocusShadowInput;
      }

      &.invalid {
        box-shadow: inset 0 -2px 0 0 $errorColor, inset 0 2px 0 0 $IconInvert,
        inset 0 10px 35px -4px $ActiveInputColor;

        &:hover, &:active {
          box-shadow: inset 0 -2px 0 0 $errorColor, inset 0 2px 0 0 $IconInvert,
          inset 0 24px 35px -4px $ActiveInputColor;
        }

        &:active {
          border: solid 1px $ActiveBorderColor;
          box-shadow: inset 0 -2px 0 0 $errorColor, inset 0 2px 0 0 $IconInvert;
        }

        &:focus {
          box-shadow: inset 0 -2px 0 0 $errorColor, inset 0 2px 0 0 $IconInvert,
          inset 0 24px 35px -4px $ActiveInputColor,
          0 0 0 4px rgba($color: $errorColor, $alpha: 0.2);
        }

        & + .inputGroupError {
          bottom: 0;
          display: block;
        }
      }
    }
  }

  &Error {
    position: absolute;
    display: none;
    font-size: 12px;
    line-height: 1.33;
    color: $errorColor;
  }
}
