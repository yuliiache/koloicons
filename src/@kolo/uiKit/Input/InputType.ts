export enum InputType {
  TEXT = 'text',
  PASSWORD = 'password',
  EMAIL = 'email',
  NUMBER = 'number',
  PHONE = 'phone',
}
