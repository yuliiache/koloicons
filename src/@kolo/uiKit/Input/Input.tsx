import classNames from 'classnames';
import React from 'react';

import styles from './Input.module.scss';
import {InputType} from './InputType';

interface Props {
  label?: string;
  placeholder?: string;
  isDisabled?: boolean;
  errorMessage?: string;
  hint?: string;
  hasAutoFocus?: boolean;
  id: string;
  type: InputType;
  value: string;
  onChange: (value: string) => void;
}

const Input: React.FC<Props> = ({
  id,
  label,
  placeholder,
  errorMessage,
  hint,
  type,
  value,
  onChange,
  isDisabled = false,
  hasAutoFocus = false,
}) => {
  const inputClasses = classNames(styles.inputGroupInput, {
    [styles.invalid]: errorMessage,
  });

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    onChange(e.target.value);
  };

  return (
    <div className={styles.inputGroup}>
      {label && (
        <label
          className={styles.inputGroupLabel}
          htmlFor={id}
        >
          {label}
        </label>
      )}
      {hint && <span className={styles.inputGroupHint}>{hint}</span>}
      <input
        className={inputClasses}
        type={type}
        id={id}
        placeholder={placeholder}
        disabled={isDisabled}
        autoFocus={hasAutoFocus}
        value={value}
        onChange={handleChange}
      />
      {errorMessage && <span className={styles.inputGroupError}>{errorMessage}</span>}
    </div>
  );
};

export default Input;
