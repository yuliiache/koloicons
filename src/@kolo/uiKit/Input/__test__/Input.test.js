import React from 'react';
import {fireEvent, render} from '@testing-library/react';

import '@testing-library/jest-dom/extend-expect';

import Input from '../Input';

const renderInputWithProps = (props) => {
  return render(<Input {...props} />);
};

describe('Input component tests', () => {
  const commonProps = {
    id: 'test',
    label: 'Test Label',
    placeholder: 'Enter text',
    type: 'text',
    hasAutoFocus: false,
    value: 'Test Value',
    onChange: jest.fn(),
  };

  const testCases = [
    {
      name: 'renders with required props',
      props: {
        id: 'test',
        type: 'text',
        value: 'value',
        onChange: jest.fn(),
      },
    },
    {
      name: 'renders with all props',
      props: {
        ...commonProps,
        isDisabled: false,
        errorMessage: 'Error message',
        hasAutoFocus: false,
        hint: 'Hint text',
      },
    },
    {
      name: 'handles input change',
      props: {...commonProps, label: 'Test Label'},
      action: (getByLabelText) => {
        const inputElement = getByLabelText('Test Label');
        fireEvent.change(inputElement, {target: {value: 'New Value'}});
      },
    },
    {
      name: 'renders as disabled',
      props: {...commonProps, isDisabled: true},
      assertion: (container) => {
        expect(container.querySelector('input')).toBeDisabled();
      },
    },
    {
      name: 'renders with invalid state',
      props: {...commonProps, errorMessage: 'Invalid input'},
      assertion: (container) => {
        expect(container.querySelector('.invalid')).toBeInTheDocument();
      },
    },
    {
      name: 'renders with hint',
      props: {...commonProps, hint: 'Hint text'},
      assertion: (container) => {
        expect(container.querySelector('.inputGroupHint')).toBeInTheDocument();
      },
    },
    {
      name: 'renders with label',
      props: {...commonProps, label: 'Test Label'},
      assertion: (container) => {
        expect(container.querySelector('.inputGroupLabel')).toBeInTheDocument();
      },
    },
    {
      name: 'renders with placeholder',
      props: {...commonProps, placeholder: 'Enter text'},
      assertion: (container) => {
        expect(container.querySelector('input')).toHaveAttribute('placeholder', 'Enter text');
      },
    },
    {
      name: 'renders with autofocus',
      props: {...commonProps, hasAutoFocus: true},
      assertion: (container) => {
        expect(container.querySelector('input')).toHaveFocus();
      },
    },
    {
      name: 'renders without autofocus',
      props: {...commonProps, hasAutoFocus: false},
      assertion: (container) => {
        expect(container.querySelector('input')).not.toHaveFocus();
      },
    },
  ];

  testCases.forEach(({name, props, action, assertion}) => {
    it(`${name}`, () => {
      const {container, getByLabelText} = renderInputWithProps(props);
      if (action) action(getByLabelText);
      if (assertion) assertion(container);
      expect(container).toMatchSnapshot();
    });
  });

  describe('Input component snapshots', () => {
    for (const isDisabled of [true, false]) {
      for (const errorMessage of [undefined, 'Error message']) {
        for (const hint of [undefined, 'Hint text']) {
          for (const hasAutoFocus of [true, false]) {
            const props = {
              ...commonProps,
              isDisabled,
              errorMessage,
              hint,
              hasAutoFocus,
            };

            it(`renders correctly - isDisabled: ${isDisabled}, errorMessage: ${errorMessage}, hint: ${hint}, hasAutoFocus: ${hasAutoFocus}`, () => {
              const {container} = renderInputWithProps(props);
              expect(container).toMatchSnapshot();
            });
          }
        }
      }
    }
  });
});
