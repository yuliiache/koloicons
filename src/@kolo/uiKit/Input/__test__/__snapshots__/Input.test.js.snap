// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component tests Input component snapshots renders correctly - isDisabled: false, errorMessage: Error message, hint: Hint text, hasAutoFocus: false 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <span
      class="inputGroupHint"
    >
      Hint text
    </span>
    <input
      class="inputGroupInput invalid"
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
    <span
      class="inputGroupError"
    >
      Error message
    </span>
  </div>
</div>
`;

exports[`Input component tests Input component snapshots renders correctly - isDisabled: false, errorMessage: Error message, hint: Hint text, hasAutoFocus: true 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <span
      class="inputGroupHint"
    >
      Hint text
    </span>
    <input
      class="inputGroupInput invalid"
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
    <span
      class="inputGroupError"
    >
      Error message
    </span>
  </div>
</div>
`;

exports[`Input component tests Input component snapshots renders correctly - isDisabled: false, errorMessage: Error message, hint: undefined, hasAutoFocus: false 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <input
      class="inputGroupInput invalid"
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
    <span
      class="inputGroupError"
    >
      Error message
    </span>
  </div>
</div>
`;

exports[`Input component tests Input component snapshots renders correctly - isDisabled: false, errorMessage: Error message, hint: undefined, hasAutoFocus: true 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <input
      class="inputGroupInput invalid"
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
    <span
      class="inputGroupError"
    >
      Error message
    </span>
  </div>
</div>
`;

exports[`Input component tests Input component snapshots renders correctly - isDisabled: false, errorMessage: undefined, hint: Hint text, hasAutoFocus: false 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <span
      class="inputGroupHint"
    >
      Hint text
    </span>
    <input
      class="inputGroupInput"
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
  </div>
</div>
`;

exports[`Input component tests Input component snapshots renders correctly - isDisabled: false, errorMessage: undefined, hint: Hint text, hasAutoFocus: true 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <span
      class="inputGroupHint"
    >
      Hint text
    </span>
    <input
      class="inputGroupInput"
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
  </div>
</div>
`;

exports[`Input component tests Input component snapshots renders correctly - isDisabled: false, errorMessage: undefined, hint: undefined, hasAutoFocus: false 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <input
      class="inputGroupInput"
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
  </div>
</div>
`;

exports[`Input component tests Input component snapshots renders correctly - isDisabled: false, errorMessage: undefined, hint: undefined, hasAutoFocus: true 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <input
      class="inputGroupInput"
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
  </div>
</div>
`;

exports[`Input component tests Input component snapshots renders correctly - isDisabled: true, errorMessage: Error message, hint: Hint text, hasAutoFocus: false 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <span
      class="inputGroupHint"
    >
      Hint text
    </span>
    <input
      class="inputGroupInput invalid"
      disabled=""
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
    <span
      class="inputGroupError"
    >
      Error message
    </span>
  </div>
</div>
`;

exports[`Input component tests Input component snapshots renders correctly - isDisabled: true, errorMessage: Error message, hint: Hint text, hasAutoFocus: true 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <span
      class="inputGroupHint"
    >
      Hint text
    </span>
    <input
      class="inputGroupInput invalid"
      disabled=""
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
    <span
      class="inputGroupError"
    >
      Error message
    </span>
  </div>
</div>
`;

exports[`Input component tests Input component snapshots renders correctly - isDisabled: true, errorMessage: Error message, hint: undefined, hasAutoFocus: false 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <input
      class="inputGroupInput invalid"
      disabled=""
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
    <span
      class="inputGroupError"
    >
      Error message
    </span>
  </div>
</div>
`;

exports[`Input component tests Input component snapshots renders correctly - isDisabled: true, errorMessage: Error message, hint: undefined, hasAutoFocus: true 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <input
      class="inputGroupInput invalid"
      disabled=""
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
    <span
      class="inputGroupError"
    >
      Error message
    </span>
  </div>
</div>
`;

exports[`Input component tests Input component snapshots renders correctly - isDisabled: true, errorMessage: undefined, hint: Hint text, hasAutoFocus: false 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <span
      class="inputGroupHint"
    >
      Hint text
    </span>
    <input
      class="inputGroupInput"
      disabled=""
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
  </div>
</div>
`;

exports[`Input component tests Input component snapshots renders correctly - isDisabled: true, errorMessage: undefined, hint: Hint text, hasAutoFocus: true 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <span
      class="inputGroupHint"
    >
      Hint text
    </span>
    <input
      class="inputGroupInput"
      disabled=""
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
  </div>
</div>
`;

exports[`Input component tests Input component snapshots renders correctly - isDisabled: true, errorMessage: undefined, hint: undefined, hasAutoFocus: false 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <input
      class="inputGroupInput"
      disabled=""
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
  </div>
</div>
`;

exports[`Input component tests Input component snapshots renders correctly - isDisabled: true, errorMessage: undefined, hint: undefined, hasAutoFocus: true 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <input
      class="inputGroupInput"
      disabled=""
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
  </div>
</div>
`;

exports[`Input component tests handles input change 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <input
      class="inputGroupInput"
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
  </div>
</div>
`;

exports[`Input component tests renders as disabled 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <input
      class="inputGroupInput"
      disabled=""
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
  </div>
</div>
`;

exports[`Input component tests renders with all props 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <span
      class="inputGroupHint"
    >
      Hint text
    </span>
    <input
      class="inputGroupInput invalid"
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
    <span
      class="inputGroupError"
    >
      Error message
    </span>
  </div>
</div>
`;

exports[`Input component tests renders with autofocus 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <input
      class="inputGroupInput"
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
  </div>
</div>
`;

exports[`Input component tests renders with hint 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <span
      class="inputGroupHint"
    >
      Hint text
    </span>
    <input
      class="inputGroupInput"
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
  </div>
</div>
`;

exports[`Input component tests renders with invalid state 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <input
      class="inputGroupInput invalid"
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
    <span
      class="inputGroupError"
    >
      Invalid input
    </span>
  </div>
</div>
`;

exports[`Input component tests renders with label 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <input
      class="inputGroupInput"
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
  </div>
</div>
`;

exports[`Input component tests renders with placeholder 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <input
      class="inputGroupInput"
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
  </div>
</div>
`;

exports[`Input component tests renders with required props 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <input
      class="inputGroupInput"
      id="test"
      type="text"
      value="value"
    />
  </div>
</div>
`;

exports[`Input component tests renders without autofocus 1`] = `
<div>
  <div
    class="inputGroup"
  >
    <label
      class="inputGroupLabel"
      for="test"
    >
      Test Label
    </label>
    <input
      class="inputGroupInput"
      id="test"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
  </div>
</div>
`;
